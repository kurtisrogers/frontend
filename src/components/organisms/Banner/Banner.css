.banner {
  position: relative;
  align-items: center;
  height: 100vh;
}
