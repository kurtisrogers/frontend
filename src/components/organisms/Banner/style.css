.banner {
  display: grid;
  width: 100%;
  grid-auto-flow: column;
  grid-auto-columns: 1fr;
  grid-auto-rows: min-content;
  /* overflow-x: clip; */
  gap: 1rem;
}

.banner--image {
  display: grid;
  width: 100%;
  min-height: calc(100vh - 53.5px);
  grid-auto-flow: column;
  grid-auto-columns: 1fr;
  grid-auto-rows: min-content;
  /* overflow-x: clip; */
  gap: 1rem;
  align-content: center;
}

.banner .banner__content {
  grid-column: span 14;
  align-self: center;
  z-index: 1;
}

.banner--image .banner__content {
  grid-column: span 6;
  align-self: center;
  z-index: 1;
}

.banner__content ul {
  padding-left: 1em;
}

.banner__content h1 {
  margin-bottom: 0px;
}

.banner__image {
  grid-column: span 8;
}

.banner--image img {
  display: block;
  object-fit: cover;
  width: 130%;
  height: calc(100vh - 107px);
  margin-left: -30%;
}