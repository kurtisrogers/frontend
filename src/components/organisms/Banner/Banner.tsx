import { Show } from 'solid-js';
import type { Image } from '@/types/branding';
import "./Banner.css";

interface Props {
  title?: string;
  backgroundImage?: Image;
};

export default function Banner( props: Readonly<Props> ) {
  const { title, backgroundImage } = props;

  if (!title) return;

  return (
    <div class="banner content-grid">
      <h1>{title}</h1>
      {/* weird image/background colour section */}
      <Show when={backgroundImage}>
        <div class="banner__image">
          <img src={backgroundImage?.src} alt={backgroundImage?.alt} />
        </div>
      </Show>
    </div>
  )
};
