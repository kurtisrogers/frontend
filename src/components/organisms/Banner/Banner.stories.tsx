import Banner from "./index";
import "./style.css";

export default {
  title: "Organisms/Banner",
  component: Banner,
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/solid/writing-docs/docs-page
  tags: ["autodocs"],
  parameters: {
    // More on how to position stories at: https://storybook.js.org/docs/solid/configure/story-layout
    layout: "fullscreen"
  }
};

const propData = {
  title: "Example",
  children: (
    <>
      <p>My name is Kurtis Rogers, a proper Software Engineer from Bristol, UK</p>
      <ul>
        <li>Specialising in Frontend technology since 2018</li>
        <li>Expert in leading frameworks including serverside Node development (Express/Koa)</li>
        <li>
          Experienced in building and developing accessible themes in both Drupal and WordPress
        </li>
        <li>Effective and approachable mentor</li>
        <li>
          Team player, able to build robust yet relaxed relationships with colleagues and
          stakeholders alike
        </li>
      </ul>
      <p>Want to know more?</p>
      <a class="button" href="/about">
        Read more about me
      </a>
    </>
  ),
  backgroundImage: {
    data: {
      attributes: {
        alternativeText:
          "A photo of The Milky Way, lots of stars and some star-making gas clouds far far away...",
        url: "/images/pexels-jack-davis-86003658-9069564-thumbnail.webp",
        formats: {
          thumbnail: {
            src: "/images/pexels-jack-davis-86003658-9069564-thumbnail.webp"
          },
          small: {
            src: "/images/pexels-jack-davis-86003658-9069564-sm.webp",
            minWidth: 320
          },
          medium: {
            src: "/images/pexels-jack-davis-86003658-9069564-md.webp",
            minWidth: 768
          },
          large: {
            src: "/images/pexels-jack-davis-86003658-9069564-lg.webp",
            minWidth: 960
          },
          xlarge: {
            src: "/images/pexels-jack-davis-86003658-9069564-xl.webp",
            minWidth: 1420
          }
        }
      }
    }
  }
};

const Template = () => {
  return (
    <Banner title={propData.title} backgroundImage={propData.backgroundImage}>
      {propData.children}
    </Banner>
  );
};

export const Default = Template.bind({});
// Default.args = {
//   // Add any props you want to pass to the Header component
// };
