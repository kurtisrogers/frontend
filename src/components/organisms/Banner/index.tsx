import { JSX, Show } from "solid-js";
import type { ImageResponse } from "@/types/branding";
import "./style.css";

import type { gridLayoutOptions } from "@/types/grid";
import responsiveImages from "@/helpers/responsiveImages";
import CSSFilterPanel from "@/components/atoms/CSSFilterPanel";
import { Dynamic } from "solid-js/web";
import { LayoutSpacingDataType } from "@/helpers/layoutSpacingHandler";
import type { Heading } from "@/types/branding";

interface Props {
  title?: Heading | string;
  backgroundImage: {
    data: ImageResponse;
  };
  gridLayout?: gridLayoutOptions;
  children?: JSX.Element;
  firstChild: boolean;
  variant: "light" | "dark";
  spacing: LayoutSpacingDataType;
}

export default function Banner(props: Readonly<Props>) {
  const { title, backgroundImage, gridLayout, children, firstChild, variant = "dark" } = props;

  return (
    <section
      class={`banner content-grid ${gridLayout ?? ""} ${firstChild ? "first-element" : "page-sibling"} content-grid--${variant}`}
    >
      <div class={`banner${!!backgroundImage ? "--image" : ""}`}>
        <div class={`banner__content`}>
          <CSSFilterPanel variant="blur">
            <Show when={title}>
              <Dynamic
                component={`h${typeof title === "string" ? "1" : title?.headingLevel}`}
                class={typeof title === "string" ? "text-size-h1" : title?.headingClass}
              >
                {typeof title === "string" ? title : title?.text}
              </Dynamic>
            </Show>
            {children}
          </CSSFilterPanel>
        </div>
        {/* weird image/background colour section - not fully made my mind up */}
        <Show when={backgroundImage}>
          {/* create an image component - generate images here: https://isrcset.com/generate */}
          <div class="banner__image">
            {/* TODO: create a picture component which accepts an array of responsive images */}
            {responsiveImages(backgroundImage)}
          </div>
        </Show>
      </div>
    </section>
  );
}
