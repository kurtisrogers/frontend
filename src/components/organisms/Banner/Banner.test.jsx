import { describe, expect, vi, beforeEach, afterEach, it } from "vitest";
import { render } from "@solidjs/testing-library";
import { screen } from '@testing-library/dom';
import Banner from '.';

describe("Banner", () => {
  beforeEach(() => {
    // Mock the fetch function
    global.fetch = vi.fn();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  const propData = {
    title: "Welcome to my website",
    children: <>
      <p>My name is Kurtis Rogers, a proper Software Engineer from Bristol, UK</p>
      <ul>
        <li>Specialising in Frontend technology since 2018</li>
        <li>Expert in leading frameworks including serverside Node development (Express/Koa)</li>
        <li>Experienced in building and developing accessible themes in both Drupal and WordPress</li>
        <li>Effective and approachable mentor</li>
        <li>Team player, able to build robust yet relaxed relationships with colleagues and stakeholders alike</li>
      </ul>
      <p>Want to know more?</p>
      <a class="button" href="/about">Read more about me</a>
    </>,
    backgroundImage: {
      data: {
        attributes: {
          alternativeText: 'A photo of The Milky Way, lots of stars and some star-making gas clouds far far away...',
          url: '/images/pexels-jack-davis-86003658-9069564-thumbnail.webp',
          formats: {
            thumbnail: {
              src: '/images/pexels-jack-davis-86003658-9069564-thumbnail.webp',
            },
            small: {
              src: '/images/pexels-jack-davis-86003658-9069564-sm.webp',
              minWidth: 320,
            },
            medium: {
              src: '/images/pexels-jack-davis-86003658-9069564-md.webp',
              minWidth: 768,
            },
            large: {
              src: '/images/pexels-jack-davis-86003658-9069564-lg.webp',
              minWidth: 960,
            },
            xlarge: {
              src: '/images/pexels-jack-davis-86003658-9069564-xl.webp',
              minWidth: 1420,
            },
          },
        },
      },
    },
    gridLayout: "wide",
  }

  it("renders", () => {
    const { container } = render(() => <Banner {...propData} />);
    expect(container).toMatchSnapshot();
  });
  
  it("has a title", () => {
    const { getByText } = render(() => <Banner {...propData} />);
    
    expect(getByText(/Welcome to my website/i)).toBeInTheDocument();
  });
  
  it("can have a image", () => {
    render(() => <Banner {...propData} />);
    
    // Find the image inside the picture element
    const image = screen.getByAltText(/A photo of The Milky Way, lots of stars and some star-making gas clouds far far away.../i);
    
    // Check if the image exists and has the correct alt text
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute("src", propData.backgroundImage.url);
  });
});