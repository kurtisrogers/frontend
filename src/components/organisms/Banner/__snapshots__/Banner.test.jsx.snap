// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Banner > renders 1`] = `
<div>
  <section
    class="banner page-sibling content-grid--dark"
  >
    <div
      class="banner__content"
    >
      <hgroup
        class="banner__content--copy fade--in_up"
      >
        <h1
          class="text-size-h2"
        >
          Software Engineer
        </h1>
        <p>
          My name is Kurtis Rogers and I'm an experienced fullstack software engineer from Bristol, UK.
        </p>
        <ul>
          <li>
            An expert in popular frontend frameworks including serverside Node development (Express/Koa)
          </li>
          <li>
            Effective and approachable mentor to less experienced colleagues, always willing to help folks where I can.
          </li>
          <li>
            A team player, able to build robust yet relaxed relationships with all stakeholders.
          </li>
        </ul>
        <ul
          class="buttongroup"
        >
          <li>
            <a
              class="bg-red text-white  "
              href="https://www.linkedin.com/in/kurtisrogers"
              target="_blank"
            >
              Visit my LinkedIn
              <span
                class="sr-only"
              >
                 (opens a new tab)
              </span>
            </a>
          </li>
          <li>
            <a
              class="bg-white text-black  "
              href="https://github.com/Kurtmcmurt/kurtisrogers.com"
              target="_blank"
            >
              Repo for the site
              <span
                class="sr-only"
              >
                 (opens a new tab)
              </span>
            </a>
          </li>
        </ul>
      </hgroup>
      <div
        class="banner__content--image fade--in"
      >
        <picture>
          <source
            media="(min-width: 1420px)"
            srcset="/images/optimised/milkyway-galaxy-forest-xl.webp"
          />
           
          <source
            media="(min-width: 960px)"
            srcset="/images/optimised/milkyway-galaxy-forest-lg.webp"
          />
           
          <img
            alt="A photo of The Milky Way, taken from the surface of Earth in a heavily wooded area. The silouhette of trees wrap the image with the sky in the background emphasising the expanse of our galaxy."
            class="w-full h-full object-cover"
            decoding="async"
            loading="lazy"
            src="/images/optimised/milkyway-galaxy-forest-md.webp"
          />
        </picture>
        <div
          class="credit"
        >
          
          <span>
            A photo of The Milky Way, taken from the surface of Earth in a heavily wooded area. The silouhette of trees wrap the image with the sky in the background emphasising the expanse of our galaxy.
          </span>
        </div>
      </div>
    </div>
  </section>
</div>
`;
